### ISSUE TITLE
Question about dimension of Hessian matrix of conv layers

### ISSUE THREAD

**Issue by adpatil234** [csyhhu/L-OBS#2]
Hi Shangyu, 

Thanks for sharing the code!

I have question about the way Hessian matrix is calculated in "calculate_hessian_conv_tf" function, which is in file calculate_hessian_inverse.py

Consider a convolutional layer consisting of 128 filters, each having dimensions 3x3x64. So, here number of input channels is 64, while number of output channels is 128. 

For this layer, Hessian matrix computed by your function "calculate_hessian_conv_tf" has dimensions 64x64. However, following your NIPS2017 paper, I think, the dimensions of the Hessian matrix in this case should be (3x3x64)x(3x3x64) = 576x576. This is because, there are 3x3x64=576 weights in the filter.  

Firthermore, when I used the 64x64 dimensional Hessian computed above to compute sensitivities, I get a dimensional mismatch error in function "prune_weights_conv". specifically, it is complaining about following line: 

<sensitivity =  (0.5 * (row_kernel ** 2) ) / diag_hess_inv   #I am ignoring biases here > 

The error is that "row_kernel" has shape (576,) while "diag_hess_inv" has shape (64,). I think, this error is because the Hessian matrix computed here is of dimensions 64x64, instead of 576x576. 

Please let me know what I am missing here and how you resolved such mismatch in your ResNet-50 simulations.

Thanks for your help.

Ameya Patil

**Comment by csyhhu** [comment]
Hi Ameya @adpatil234 ,

Thanks for your question.
Your understanding is correct. The dimensions of the Hessian matrix should be filters's width * height * in_channels. I test my code in "res2a_branch1" and the dimension is matched. 
Could you please tell me which layer causes such error?

Looking for your reply~

Shangyu

**Comment by adpatil234** [comment]
Hi Shangyu, 

Thanks a lot for your prompt reply. I really appreciate your help. 

It is good to know that Hessian matrix should be of dimensions filters's width * height * in_channels. It means that I am making some mistake in the way I am using Hessian computation function.  

Currently I am not really running ResNet-50. I am running a VGGNet-inspired architecture for CIFAR-10. I am using your Hessian computation and layer pruning functions for this network in my code. 

Here is the information, which, I think, will enable you to help me spot my mistake. 

Continuing my conv layer example above, the layer contains 128 filters, each having dimensions 3x3x64. So, here number of input channels is 64, while number of output channels is 128. 

Dimensions of input activations to this layer are: (no_of_batch_elements)x10x10x64. 

As per my understanding, function "calculate_hessian_conv_tf" computes the Hessian matrix for a convolutional layer. 

Here are dimensions observed after every step of computation in function calculate_hessian_conv_tf: 

`def calculate_hessian_conv_tf(layer_inputs)`
`# dimensions of layer_inputs = (no_of_batch_elements)x10x10x64`

`a = tf.expand_dims(layer_inputs, axis=-1)`
`# dimensions of a = (no_of_batch_elements)x10x10x64x1`

`b = tf.expand_dims(layer_inputs, axis=3)`
`# dimensions of b = (no_of_batch_elements)x10x10x1x64`

`outprod = tf.multiply(a, b)`
`# dimensions of outprod = (no_of_batch_elements)x10x10x64x64`

`return tf.reduce_mean(outprod, axis=[0, 1, 2])`
`# dimensions of Hessian returned here = 64x64`

I have removed two `tf.concat` operations on `a` and `b` in your original function, since I am not worrying about biases for now. 

As we discussed above, this function should actually return Hessian having dimensions (3x3x64)x(3x3x64)=576x576. But as per my code above, it only returns a matrix of dimensions 64x64. 

Please let me know what I am missing here. Is the format of layer_inputs incorrect, or is this not the final Hessian for conv layer? As per my understanding, the function "calculate_hessian_conv_tf" doesn't ask for filter_width and filter_height as its inputs. 

Again, thank you very much for your help. I look forward to your reply. 

Ameya

**Comment by csyhhu** [comment]
Hi Ameya @adpatil234 
Thanks for your detailed explanations. 
The problem seems to be the dimensions of layer inputs, I think '10' should be the input features (last layer's output)' width and height. However, the actual input features dimensions used in Hessian calculation should be 3*3*64, so what you get from "dimensions of layer inputs" should be: (no_of_batch_elements) x a x b x (3*3*64), where a and b represent the output dimension of this layers. Here is my result in my computers:

``In [1]: run calculate_hessian_inverse.py
Layer conv1, input shape: (?, 112, 112, 147)
``

In conv1, the input shape is (?, 112, 112, 147), 112 means the output width and height of this layer, 147 = 7*7*3, that is because in conv layer, its kernel dimension is (7,7, 3, 64), where 64 is the output channel.

The difference between your result and mine is that you may miss the function tf.extract_image_patches, which extract patches from input features and store them as 3*3*64 (for example).  

To summary, in convolution layer, "layer input" refers to the extracted patches instead of the output of last layer, which is also a keypoint in understanding L-OBS. 

However, the most weird thing is that due to that code is written long time ago, I can't find the extract_image_patches function even in my code, but it still comes up with the desired results. I am also very confused about that. Currently I generated Hessian matrix using PyTorch. I will upload a new version of L-OBS using PyTorch, which will be more straightforward to understand the code.

Please let me know if these make sense to you. 

Shangyu

**Comment by adpatil2** [comment]
Hello Shangyu, 

Sorry for really late response here. 

Your response above made perfect sense and did help a lot. 

Yes, I was indeed missing tf.extract_image_patches function. After your response, I could locate it in your Tensorflow code as well. The Hessian calculations then worked correctly as per your explanation above. 

Thanks for uploading PyTorch implementation. As you had mentioned, it is definitely much easier and straightforward to understand it. I have came across it just recently. 

Thank you very much for your help. I am closing this issue. 

Ameya
